src/pages/[slug]/p.tsx

```diff
@@ -1,45 +1,59 @@
 import { useSession } from '@faststore/sdk'
+import { gql } from '@vtex/graphql-utils'
 import { graphql } from 'gatsby'
 import {
   BreadcrumbJsonLd,
   GatsbySeo,
   ProductJsonLd,
 } from 'gatsby-plugin-next-seo'
-import React from 'react'
+import { useEffect } from 'react'
 import ProductDetails from 'src/components/sections/ProductDetails'
 import ProductShelf from 'src/components/sections/ProductShelf'
 import { mark } from 'src/sdk/tests/mark'
 import type { PageProps } from 'gatsby'
 import type {
   ProductPageQueryQuery,
+  ServerProductPageQueryQuery,
   ProductPageQueryQueryVariables,
 } from '@generated/graphql'
 import { ITEMS_PER_SECTION } from 'src/constants'
 
 export type Props = PageProps<
   ProductPageQueryQuery,
-  ProductPageQueryQueryVariables
->
+  ProductPageQueryQueryVariables,
+  unknown,
+  ServerProductPageQueryQuery
+> & { slug: string }
 
 function Page(props: Props) {
   const { locale, currency } = useSession()
   const {
-    data: { product, site },
+    data: { site },
+    serverData: { product },
     location: { host },
-    params: { slug },
+    slug,
   } = props
 
-  if (!product) {
-    throw new Error('NotFound')
-  }
-
-  const title = product.seo.title ?? site?.siteMetadata?.title ?? ''
+  const notFound = !product
+  const title = product?.seo.title ?? site?.siteMetadata?.title ?? ''
   const description =
-    product.seo.description ?? site?.siteMetadata?.description ?? ''
+    product?.seo.description ?? site?.siteMetadata?.description ?? ''
 
   const canonical =
     host !== undefined ? `https://${host}/${slug}/p` : `/${slug}/p`
 
+  useEffect(() => {
+    if (notFound) {
+      window.location.href = `/404/?from=${encodeURIComponent(
+        window.location.pathname
+      )}`
+    }
+  }, [notFound])
+
+  if (notFound) {
+    return null
+  }
```

Comment:
just thinking outloud: is it possible to redirect from the `getServerData` instead?

Reply:
Yes, we can do it later on